text/0000-macro-hygiene-optout.md

```diff
@@ -0,0 +1,229 @@
+- It is not immediately obvious from a macro definition which (occurences of) identifiers take their syntax context from the call site. One has to read through the whole definition to figure it out.
+- The syntax permits marking identifiers with call-site hygiene purely for "use" or "import" scenarios (as opposed to "defining" or "exporting" scenarios). Parameters are intended for this purpose, and accomplish the task much better, since they self-document uses of identifiers. However, this ability may actually be desirable more than problematic, as mentioned in the [usage notes][usage-notes].
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The design in this RFC was chosen because of its simple syntax and semantics, and the fact it offers a good way to get experience with hygiene opt-out in general, due to its fine-grainedness.
+
+The main alternative considered was having an `escapes` attribute for macros and not using a sigil.
+
+```rust
+#[escapes(S, T)]
+macro m() {
+    struct S; // Defines `S` at the call-site.
+    T // Resolves at the call-site.
+}
+```
+
+The above would then be equivalent to the following, using the sigil syntax.
+
+```rust
+macro m() {
+    struct #S; // Defines `S` at the call-site.
+    #T // Resolves at the call-site.
+}
+```
+
+The obvious benefit of this is that is manifest which identifiers (`S` and `T` in the above example) are hygiene-escaped. A downside, which may or may not be significant, is that these identifiers are then *always* escaped within the macro definition, and thus can never be used with definition-site hygiene.
+
+Going beyond a single `escapes` attribute, one can also imagine having two separate attributes: `defines`, for defining (exporting) identifiers, and `uses`, for using (importing) identifiers. The main issue here is the complexity of the semantics and implementation; indeed, it is not even clear whether one could clearly demarcate cases of definition and use at the syntactical level. As implied by the [usage notes][usage-notes], however, the `uses` attribute would largely overlap with the purpose of macro parameters.
+
+In the end, the approach taken by this RFC was chosen due to the fact it has the most prior art, including an [existing working implementation][pr-47992]. It is also the most flexible in that it allows different hygiene to be applied to different *occurrences* of the same identifier. This will allow us to learn more about the use of hygiene opt-out in practice, while the feature is unstable.
+
+# Prior art
+[prior-art]: #prior-art
+
+Extended discussion on this subject was carried out in a [pull request][pr-47992] for this feature, which was closed due to the decision that an RFC such as this one be accepted first. [Alternatives][pr-47992-alternatives] were originally evaluated there, with discussion initiated by @jseyfried, and [continued][pr-47992-alternatives-eval] by @petrochenkov.
```

Comment:
I'd expect some discussion of how this works in other languages here. In particular, Scheme has a rich system for doing this sort of thing.

Reply:
Hmm. I'd like to avoid learning Scheme properly for this... maybe I can dig up a decent explanation somewhere?